### PR TITLE
customize object mapper from the object mapper factory

### PR DESCRIPTION
**Pull request by kxmas**
Address https://github.com/microsoft/spring-data-cosmosdb/issues/380 and maintains ZonedDateTimeDeserializer customization.

### CODE REVIEW

**Comment by Incarnation-p-lee** [comment]
Could you please fix some CI reported issue ? So we can start the review process.

**Comment by kxmas** [comment]
> Could you please fix some CI reported issue ? So we can start the review process.

I cleaned up the imports, but I'm not sure what to do about the failing CI build. This is a minor change that didn't affect the connection settings

**Comment by kushagraThapar** [comment]
@kxmas The problem with customizing object mapper in ObjectMapperFactory is in case spring instantiates the object mapper bean externally, you loose all the customizations. 

In the latest master, you will see, I have now customized objectMapper bean in the constructor of MappingDocumentDbConverter. That seems to be working fine in both the cases. (when spring instantiates, or we instantiate from ObjectMapperFactory. If you agree, please close this PR.

**Comment by kxmas** [comment]
hi @kushagraThapar,

It's still broken in the sense that if I instantiate my own objectMapper bean, the constructor in MappingDocumentDbConverter is registering a converter that could possibly override the converter that I already configured, along with other settings.

**Comment by kushagraThapar** [comment]
@kxmas , How are you specifying the object mapper?
Are you using @Qualifier(Constants.OBJECTMAPPER_BEAN_NAME) property to instantiate your mapper and then relying on spring to autowire it ?

**Comment by kxmas** [comment]
> @kxmas , How are you specifying the object mapper?
> Are you using @qualifier(Constants.OBJECTMAPPER_BEAN_NAME) property to instantiate your mapper and then relying on spring to autowire it ?
> 
> If that's the case, I think I understand your issue now, and it makes sense to me.

Yes, I was creating a objectmapper with that bean name and letting spring autowire it.

I'm glad it makes sense now :)

File: src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/convert/ObjectMapperFactory.java [hunk 1]
```diff
@@ -6,22 +6,37 @@
...
 public class ObjectMapperFactory {
     private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
 
     static {
+        OBJECT_MAPPER.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
         OBJECT_MAPPER.registerModule(new ParameterNamesModule())
                         .registerModule(new Jdk8Module())
                         .registerModule(new JavaTimeModule());
+
+        if (!OBJECT_MAPPER.canDeserialize(OBJECT_MAPPER.constructType(ZonedDateTime.class))) {
```
Comment (kushagraThapar) [inline src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/convert/ObjectMapperFactory.java]: Is this just a safety catch or a convention when adding modules ?

**Comment by kushagraThapar** [comment]
@kxmas thanks for explanation. It makes sense now. I just started owning this repo, so still exploring it :) 
Your PR makes sense. I just have one question regarding the `canDeserialize` call.

**Comment by kxmas** [comment]
Yeah, I can probably find some time to that.

**Comment by kushagraThapar** [comment]
> Yeah, I can probably find some time to that.

Thanks!

File: src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/convert/ObjectMapperFactory.java (hunk 1 above)
Comment (kxmas) [reply src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/convert/ObjectMapperFactory.java]: The JavaTimeModule should covert this type. I'll remove it.

**Comment by kxmas** [comment]
Tests have been updated.

**Comment by kushagraThapar** [comment]
> Tests have been updated.

Thank you @kxmas for updating the tests, I looked at the code and it looks good. 
One question I have though is, as you mentioned that `JavaTimeModule` should handle the ZonedDateTimeDeserializer, why remove the ZonedDataTimeDeserializerTest ?
Shouldn't we keep that test and update it to use Java Time Module ?

**Comment by kxmas** [comment]
> > Tests have been updated.
> 
> Thank you @kxmas for updating the tests, I looked at the code and it looks good.
> One question I have though is, as you mentioned that `JavaTimeModule` should handle the ZonedDateTimeDeserializer, why remove the ZonedDataTimeDeserializerTest ?
> Shouldn't we keep that test and update it to use Java Time Module ?

No, the tests were just exercising the deserializer code that was also deleted. It doesn't make sense to test Jackson's deserialization code in this project. The Jackson2 repo is the place for that.

There's a case to be made to adding a ZonedDateTime to the integration tests though.

**Comment by kushagraThapar** [comment]
> > > Tests have been updated.
> > 
> > 
> > Thank you @kxmas for updating the tests, I looked at the code and it looks good.
> > One question I have though is, as you mentioned that `JavaTimeModule` should handle the ZonedDateTimeDeserializer, why remove the ZonedDataTimeDeserializerTest ?
> > Shouldn't we keep that test and update it to use Java Time Module ?
> 
> No, the tests were just exercising the deserializer code that was also deleted. It doesn't make sense to test Jackson's deserialization code in this project. The Jackson2 repo is the place for that.
> 
> There's a case to be made to adding a ZonedDateTime to the integration tests though.

I see, I am just skeptical about JavaTimeModule handling the zonedDateTime deserialization. Can we please add just one test case for that in this PR?
I can add more tests later.

**Comment by kxmas** [comment]
> I see, I am just skeptical about JavaTimeModule handling the zonedDateTime deserialization. Can we please add just one test case for that in this PR?
> I can add more tests later.

Why would you be a skeptic if you can [see the code](https://github.com/FasterXML/jackson-modules-java8/blob/2.9/datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/JavaTimeModule.java#L177) :)

Anyway, I updated the ZonedDateTimeDeserializerTest.

**Comment by kushagraThapar** [comment]
> > I see, I am just skeptical about JavaTimeModule handling the zonedDateTime deserialization. Can we please add just one test case for that in this PR?
> > I can add more tests later.
> 
> Why would you be a skeptic if you can [see the code](https://github.com/FasterXML/jackson-modules-java8/blob/2.9/datetime/src/main/java/com/fasterxml/jackson/datatype/jsr310/JavaTimeModule.java#L143)? :)
> 
> Anyway, I updated the ZonedDateTimeDeserializerTest.

lol, completely my bad. I apologize for my negligence. 

Thank you for taking extra time to update the tests, I will get this in after running additional tests.

**Pull request merged**